From 4d8e4e643f2ab69f9905a7c23e464df4071012cc Mon Sep 17 00:00:00 2001
From: Colm O Heigeartaigh <coheigea@apache.org>
Date: Fri, 22 Jul 2011 15:18:08 +0000
Subject: [PATCH] [SANTUARIO-284] - ElementProxy#getTextFromChildElement()
 doesn't get all of the text if the element contains an entity like &amp;

git-svn-id: https://svn.apache.org/repos/asf/santuario/xml-security-java/branches/1.4.x-fixes@1149630 13f79535-47bb-0310-9956-ffa450edef68
---
 CHANGELOG.txt                                 |  1 +
 .../xml/security/utils/ElementProxy.java      | 25 ++++++++++++-------
 2 files changed, 17 insertions(+), 9 deletions(-)

diff --git a/CHANGELOG.txt b/CHANGELOG.txt
index 60f5b4937b..bad886e5c3 100644
--- a/CHANGELOG.txt
+++ b/CHANGELOG.txt
@@ -1,6 +1,7 @@
 Changelog for "Apache xml-security" <http://santuario.apache.org/>
 
 New in v.1.4.6-SNAPSHOT:
+    Fixed SANTUARIO-284: ElementProxy#getTextFromChildElement() doesn't get all of the text if the element contains an entity like &amp;
     Fixed SANTUARIO-283: JSR105 does not retain namespace definitions on Object element when unmarshalling
     Fixed SANTUARIO-281: Invalid signature value when using XMLSignature in two EJB modules on AS
     Fixed SANTUARIO-102: Private keys must be instance of RSAPrivate or have PKCS#8 encoding
diff --git a/src/org/apache/xml/security/utils/ElementProxy.java b/src/org/apache/xml/security/utils/ElementProxy.java
index ab0d65a4de..46ad5a765f 100644
--- a/src/org/apache/xml/security/utils/ElementProxy.java
+++ b/src/org/apache/xml/security/utils/ElementProxy.java
@@ -395,15 +395,22 @@ public byte[] getBytesFromChildElement(String localname, String namespace)
     * @return the Text of the textNode
     */
    public String getTextFromChildElement(String localname, String namespace) {
-              
-         Text t =
-             (Text) XMLUtils.selectNode(
-                        this._constructionElement.getFirstChild(),
-                        namespace,
-                        localname,
-                        0).getFirstChild();
-
-         return t.getData();      
+       Node textParent = 
+           XMLUtils.selectNode(
+                   this._constructionElement.getFirstChild(),
+                   namespace,
+                   localname,
+                   0);
+       Node textChild = textParent.getFirstChild();
+       StringBuffer text = new StringBuffer();
+       while (textChild != null) {
+           if (textChild instanceof Text) {
+               text.append(((Text)textChild).getData());
+           }
+           textChild = textChild.getNextSibling();
+       }
+       
+       return text.toString(); 
    }
 
    /**
